File: swift/Sources/EHBP/Identity.swift
import Foundation
import Crypto

/// Context retained after encrypting a request, used to derive response keys (SPEC Section 4.4)
public struct RequestContext: Sendable {
    /// HPKE sender for exporting the shared secret
    public let sender: HPKE.Sender

    /// Encapsulated key sent in Ehbp-Encapsulated-Key header
    public let requestEnc: Data

    public init(sender: HPKE.Sender, requestEnc: Data) {
        self.sender = sender
        self.requestEnc = requestEnc
    }
}

/// Serializable token for decrypting a response without a live HPKE context
public struct SessionRecoveryToken: Sendable, Codable {
    public let exportedSecret: Data
    public let requestEnc: Data

    public init(exportedSecret: Data, requestEnc: Data) {
        self.exportedSecret = exportedSecret
        self.requestEnc = requestEnc
    }

    enum CodingKeys: String, CodingKey {
        case exportedSecret
        case requestEnc
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(exportedSecret.hexString, forKey: .exportedSecret)
        try container.encode(requestEnc.hexString, forKey: .requestEnc)
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        let exportedSecretHex = try container.decode(String.self, forKey: .exportedSecret)
        guard let exportedSecretData = Data(hexString: exportedSecretHex),
              exportedSecretData.count == EHBPConstants.exportLength else {
            throw DecodingError.dataCorruptedError(
                forKey: .exportedSecret,
                in: container,
                debugDescription: "invalid exportedSecret hex"
            )
        }
        self.exportedSecret = exportedSecretData

        let requestEncHex = try container.decode(String.self, forKey: .requestEnc)
        guard let requestEncData = Data(hexString: requestEncHex),
              requestEncData.count == EHBPConstants.requestEncLength else {
            throw DecodingError.dataCorruptedError(
                forKey: .requestEnc,
                in: container,
                debugDescription: "invalid requestEnc hex"
            )
        }
        self.requestEnc = requestEncData
    }
}

/// Extracts a session recovery token from an HPKE request context
public func extractSessionRecoveryToken(context: RequestContext) throws -> SessionRecoveryToken {
    let exportLabel = Data(EHBPConstants.exportLabel.utf8)
    let exportedSecret = try context.sender.exportSecret(
        context: exportLabel,
        outputByteCount: EHBPConstants.exportLength
    )

    return SessionRecoveryToken(
        exportedSecret: exportedSecret.withUnsafeBytes { Data($0) },
        requestEnc: context.requestEnc
    )
}

/// Decrypts a response body using a session recovery token
public func decryptResponseBody(
    token: SessionRecoveryToken,
    responseNonce: Data,
    encryptedData: Data
) throws -> Data {
    let keyMaterial = try deriveResponseKeys(
        exportedSecret: token.exportedSecret,
        requestEnc: token.requestEnc,
        responseNonce: responseNonce
    )

    var result = Data()
    var offset = 0
    var seq: UInt64 = 0

    while offset + 4 <= encryptedData.count {
        let chunkLength = Int(encryptedData[offset]) << 24 |
                          Int(encryptedData[offset + 1]) << 16 |
                          Int(encryptedData[offset + 2]) << 8 |
                          Int(encryptedData[offset + 3])
        offset += 4

        if chunkLength == 0 {
            continue
        }

        guard offset + chunkLength <= encryptedData.count else {
            throw EHBPError.invalidResponse("incomplete chunk at offset \(offset)")
        }

        let ciphertext = encryptedData.subdata(in: offset..<(offset + chunkLength))
        offset += chunkLength

        let plaintext = try decryptChunk(
            keyMaterial: keyMaterial,
            seq: seq,
            ciphertext: ciphertext
        )
        seq += 1

        result.append(plaintext)
    }

    return result
}

/// Client identity for EHBP encryption/decryption (SPEC Section 5.1)
public final class Identity: Sendable {
    private let publicKey: Curve25519.KeyAgreement.PublicKey
    private let ciphersuite: HPKE.Ciphersuite

    /// Creates an Identity from a server's raw public key bytes
    ///
    /// - Parameter publicKeyBytes: 32-byte X25519 public key
    public init(publicKeyBytes: Data) throws {
        guard publicKeyBytes.count == 32 else {
            throw EHBPError.invalidInput("public key must be 32 bytes, got \(publicKeyBytes.count)")
        }

        self.publicKey = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: publicKeyBytes)
        self.ciphersuite = HPKE.Ciphersuite(
            kem: .Curve25519_HKDF_SHA256,
            kdf: .HKDF_SHA256,
            aead: .AES_GCM_256
        )
    }

    /// Creates an Identity from a hex-encoded public key string.
    ///
    /// This is used by clients who already have the server's public key
    /// and don't need to fetch it.
    ///
    /// - Parameter publicKeyHex: 64-character hex string representing a 32-byte X25519 public key
    public convenience init(publicKeyHex: String) throws {
        guard let publicKeyBytes = Data(hexString: publicKeyHex) else {
            throw EHBPError.invalidInput("invalid hex string")
        }
        try self.init(publicKeyBytes: publicKeyBytes)
    }

    /// Creates an Identity from an RFC 9458 key configuration
    ///
    /// Format:
    /// - Key ID (1 byte)
    /// - KEM ID (2 bytes, big-endian)
    /// - Public Key (32 bytes for X25519)
    /// - Cipher Suites Length (2 bytes, big-endian)
    /// - For each cipher suite:
    ///   - KDF ID (2 bytes)
    ///   - AEAD ID (2 bytes)
    ///
    /// - Parameter config: RFC 9458 key configuration data
    public init(config: Data) throws {
        guard config.count >= 7 else {
            throw EHBPError.invalidInput("config too short")
        }

        var offset = 0

        // Key ID (1 byte) - skip
        offset += 1

        // KEM ID (2 bytes, big-endian)
        let kemId = UInt16(config[offset]) << 8 | UInt16(config[offset + 1])
        offset += 2

        guard kemId == HPKEConfig.kem else {
            throw EHBPError.invalidInput("unsupported KEM: 0x\(String(kemId, radix: 16))")
        }

        // Public Key (32 bytes for X25519)
        let publicKeySize = 32
        guard config.count >= offset + publicKeySize else {
            throw EHBPError.invalidInput("config too short for public key")
        }

        let publicKeyBytes = config.subdata(in: offset..<(offset + publicKeySize))
        offset += publicKeySize

        // Cipher Suites Length (2 bytes)
        guard config.count >= offset + 2 else {
            throw EHBPError.invalidInput("config too short for cipher suites length")
        }

        let cipherSuitesLength = Int(config[offset]) << 8 | Int(config[offset + 1])
        offset += 2

        // Parse first cipher suite
        guard cipherSuitesLength >= 4, config.count >= offset + 4 else {
            throw EHBPError.invalidInput("no cipher suites in config")
        }

        let kdfId = UInt16(config[offset]) << 8 | UInt16(config[offset + 1])
        let aeadId = UInt16(config[offset + 2]) << 8 | UInt16(config[offset + 3])

        guard kdfId == HPKEConfig.kdf else {
            throw EHBPError.invalidInput("unsupported KDF: 0x\(String(kdfId, radix: 16))")
        }
        guard aeadId == HPKEConfig.aead else {
            throw EHBPError.invalidInput("unsupported AEAD: 0x\(String(aeadId, radix: 16))")
        }

        self.publicKey = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: publicKeyBytes)
        self.ciphersuite = HPKE.Ciphersuite(
            kem: .Curve25519_HKDF_SHA256,
            kdf: .HKDF_SHA256,
            aead: .AES_GCM_256
        )
    }

    /// Encrypts request body and returns context for response decryption (SPEC Section 5.1)
    ///
    /// - Parameter body: Request body to encrypt
    /// - Returns: Encrypted body with chunk framing, and context needed to decrypt the response
    public func encryptRequest(body: Data) throws -> (encryptedBody: Data, context: RequestContext) {
        let info = Data(EHBPConstants.hpkeRequestInfo.utf8)

        var sender = try HPKE.Sender(
            recipientKey: publicKey,
            ciphersuite: ciphersuite,
            info: info
        )

        let encapsulatedKey = sender.encapsulatedKey
        let encrypted = try sender.seal(body)

        // Frame as: LEN (4 bytes big-endian) || ciphertext
        var chunkedData = Data()
        var length = UInt32(encrypted.count).bigEndian
        chunkedData.append(Data(bytes: &length, count: 4))
        chunkedData.append(encrypted)

        let context = RequestContext(
            sender: sender,
            requestEnc: encapsulatedKey
        )

        return (chunkedData, context)
    }

    /// Derives response decryption keys using OHTTP-style derivation (SPEC Section 4.4.1)
    ///
    /// - Parameters:
    ///   - context: Request context from encryptRequest
    ///   - responseNonce: 32 bytes from Ehbp-Response-Nonce header
    /// - Returns: Key material for decrypting response chunks
    public func deriveResponseKeys(
        context: RequestContext,
        responseNonce: Data
    ) throws -> ResponseKeyMaterial {
        let token = try extractSessionRecoveryToken(context: context)

        return try EHBP.deriveResponseKeys(
            exportedSecret: token.exportedSecret,
            requestEnc: token.requestEnc,
            responseNonce: responseNonce
        )
    }
}

File: python/src/ehbp/identity.py
"""Server identity: RFC 9458 key-config handling and HPKE request encryption."""

from __future__ import annotations

import struct
from dataclasses import dataclass

from pyhpke import AEADId, CipherSuite, KDFId, KEMId

from .derive import frame_chunk
from .errors import HPKEError, InvalidConfigError
from .protocol import (
    AEAD_AES_256_GCM,
    EXPORT_LABEL,
    EXPORT_LENGTH,
    HPKE_REQUEST_INFO,
    KDF_HKDF_SHA256,
    KEM_X25519_HKDF_SHA256,
    KEY_ID,
    REQUEST_ENC_LENGTH,
)
from .session import SessionRecoveryToken

_CIPHER_SUITE_ENTRY_SIZE = 4
_MAX_KEY_ID = 0xFF


def _new_suite() -> CipherSuite:
    return CipherSuite.new(
        KEMId.DHKEM_X25519_HKDF_SHA256, KDFId.HKDF_SHA256, AEADId.AES256_GCM
    )


@dataclass(frozen=True)
class EncryptedRequest:
    encapsulated_key: bytes
    body: bytes
    token: SessionRecoveryToken


def _read_u16(data: bytes, offset: int, field: str) -> tuple[int, int]:
    if len(data) - offset < 2:
        raise InvalidConfigError(f"missing {field}")
    return struct.unpack_from(">H", data, offset)[0], offset + 2


class ServerIdentity:
    """The server's public HPKE configuration.

    Holds only the public key; it can encrypt requests but never decrypt them.
    """

    def __init__(self, public_key: bytes, key_id: int = KEY_ID) -> None:
        if len(public_key) != REQUEST_ENC_LENGTH:
            raise InvalidConfigError(
                f"public key must be {REQUEST_ENC_LENGTH} bytes, got {len(public_key)}"
            )
        self._suite = _new_suite()
        try:
            self._public_key = self._suite.kem.deserialize_public_key(bytes(public_key))
        except Exception as err:  # noqa: BLE001 - pyhpke raises library-specific errors
            raise InvalidConfigError(f"invalid X25519 public key: {err}") from err
        self._public_key_bytes = bytes(public_key)
        if not 0 <= key_id <= _MAX_KEY_ID:
            raise InvalidConfigError(f"key id must be between 0 and {_MAX_KEY_ID}, got {key_id}")
        self._key_id = key_id

    @classmethod
    def from_public_key_bytes(cls, public_key: bytes) -> ServerIdentity:
        return cls(bytes(public_key))

    @classmethod
    def from_public_key_hex(cls, public_key_hex: str) -> ServerIdentity:
        try:
            raw = bytes.fromhex(public_key_hex)
        except ValueError as err:
            raise InvalidConfigError(f"invalid public key hex: {err}") from err
        return cls(raw)

    @classmethod
    def unmarshal_public_config(cls, data: bytes) -> ServerIdentity:
        if len(data) < 1:
            raise InvalidConfigError("missing key id")
        key_id = data[0]
        offset = 1

        kem_id, offset = _read_u16(data, offset, "KEM id")
        if kem_id != KEM_X25519_HKDF_SHA256:
            raise InvalidConfigError(f"unsupported KEM: 0x{kem_id:04x}")

        public_key_end = offset + REQUEST_ENC_LENGTH
        if public_key_end > len(data):
            raise InvalidConfigError("truncated public key")
        public_key = data[offset:public_key_end]
        offset = public_key_end

        suites_len, offset = _read_u16(data, offset, "cipher suites length")
        if suites_len == 0:
            raise InvalidConfigError("no cipher suites found in config")
        if suites_len % _CIPHER_SUITE_ENTRY_SIZE != 0:
            raise InvalidConfigError("cipher suites length must be a multiple of 4")
        if offset + suites_len > len(data):
            raise InvalidConfigError("truncated cipher suites")

        kdf_id, offset = _read_u16(data, offset, "KDF id")
        aead_id, offset = _read_u16(data, offset, "AEAD id")
        if kdf_id != KDF_HKDF_SHA256 or aead_id != AEAD_AES_256_GCM:
            raise InvalidConfigError(
                f"unsupported cipher suite: KDF=0x{kdf_id:04x}, AEAD=0x{aead_id:04x}"
            )

        return cls(public_key, key_id)

    def marshal_public_config(self) -> bytes:
        out = bytearray()
        out.append(self._key_id)
        out += struct.pack(">H", KEM_X25519_HKDF_SHA256)
        out += self._public_key_bytes
        out += struct.pack(">H", _CIPHER_SUITE_ENTRY_SIZE)
        out += struct.pack(">H", KDF_HKDF_SHA256)
        out += struct.pack(">H", AEAD_AES_256_GCM)
        return bytes(out)

    @property
    def key_id(self) -> int:
        return self._key_id

    def public_key_bytes(self) -> bytes:
        return self._public_key_bytes

    def public_key_hex(self) -> str:
        return self._public_key_bytes.hex()

    def encrypt_request_body(self, plaintext: bytes):
        """Seal a request body to the server's public key.

        Returns ``None`` for empty bodies: bodyless requests pass through
        unencrypted and receive a plaintext response (SPEC Section 7.4).
        """
        if len(plaintext) == 0:
            return None
        try:
            enc, sender = self._suite.create_sender_context(
                self._public_key, info=HPKE_REQUEST_INFO
            )
            ciphertext = sender.seal(bytes(plaintext), b"")
            exported_secret = sender.export(EXPORT_LABEL, EXPORT_LENGTH)
        except Exception as err:  # noqa: BLE001 - normalize HPKE library failures
            raise HPKEError(f"failed to encrypt request body: {err}") from err

        token = SessionRecoveryToken(exported_secret, enc)
        return EncryptedRequest(
            encapsulated_key=bytes(enc), body=frame_chunk(ciphertext), token=token
        )
